##  Measures of Central Tendency



# 1. Mean

### Definition

Mean is the average of all values in a dataset. It is calculated by adding all the values and dividing by the total number of values.

### Real-life Example

If a student scores 70, 80, and 90 in three tests, the mean gives the student's average score.

### Formula

$$
\text{Mean} = \frac{\sum x_i}{n}
$$

Where:
- $x_i$ = each value
- $n$ = total number of values

### Numerical Example

Data: 10, 20, 30, 40, 50

$$
\text{Mean} = \frac{10 + 20 + 30 + 40 + 50}{5} = 30
$$


In [1]:
import numpy as np

data = [10, 20, 30, 40, 50]

mean = np.mean(data)

print("Mean:", mean)


Mean: 30.0


### AI/ML Example

Mean can be used to fill missing numerical values in a dataset.


In [3]:
import pandas as pd

df = pd.DataFrame({
    "Age": [20, 25, None, 30, 25]
})

df["Age"] = df["Age"].fillna(df["Age"].mean())

print(df)


    Age
0  20.0
1  25.0
2  25.0
3  30.0
4  25.0


### Explanation

The code first calculates the average of the available ages and then replaces the missing (`NaN`) value with that mean.

For example, if the mean age is 25, the missing age will be replaced with 25

This is useful because many machine learning algorithms require complete numerical data. Mean imputation lets us keep the row instead of deleting it and prepare the data for model training.


# 2. Median

### Definition

Median is the middle value of a dataset after arranging the values in ascending or descending order

### Real-life Example

Suppose five employees have salaries:

20,000, ₹25,000, ₹30,000, ₹35,000, ₹2,00,000

The median salary is 30,000 because it is the middle value.

### Formula

For an odd number of observations:

$$
\text{Median} = \text{Middle Value}
$$

For an even number of observations:

$$
\text{Median} =
\frac{\text{First Middle Value} + \text{Second Middle Value}}{2}
$$

### Numerical Example

Data: 10, 20, 30, 40, 100

$$
\text{Median} = 30
$$

For an even dataset: 10, 20, 30, 40

$$
\text{Median} = \frac{20 + 30}{2} = 25
$$

### Python Implementation


In [5]:
import numpy as np

data = [10, 20, 30, 40, 100]

median = np.median(data)

print("Median:", median)


Median: 30.0


### AI/ML Example

Median can be used to fill missing numerical values when the data contains outliers


In [7]:
import pandas as pd

df = pd.DataFrame({
    "Salary": [25000, 30000, None, 35000, 200000]
})

df["Salary"] = df["Salary"].fillna(df["Salary"].median())

print(df)


     Salary
0   25000.0
1   30000.0
2   32500.0
3   35000.0
4  200000.0


### Explanation

The code calculates the median salary and replaces the missing salary with that value.

Median is especially useful when the dataset contains extreme values or outliers Here, ₹2,00,000 is much larger than the other salaries, so using the median avoids letting that extreme value strongly influence the replacement value.

This helps prepare numerical data containing outliers for machine learning.


# 3. Mode

### Definition

Mode is the value that occurs most frequently in a dataset

### Real-life Example

A shoe shop sells these sizes:

7, 8, 8, 9, 8, 10

Size 8 occurs most frequently, so the mode is 8

### Formula

$$
\text{Mode} = \text{Value with the highest frequency}
$$

### Numerical Example

Data 1, 2, 2, 2, 3, 4, 5

$$
\text{Frequency of 2} = 3
$$

$$
\text{Mode} = 2
$$

### Python Implementation


In [9]:
from scipy import stats

data = [1, 2, 2, 2, 3, 4, 5]

mode = stats.mode(data)

print("Mode:", mode.mode)


Mode: 2


### AI/ML Example

Mode is commonly used to fill missing categorical values


In [11]:
import pandas as pd

df = pd.DataFrame({
    "City": ["Bengaluru", "Mysuru", "Bengaluru", None, "Bengaluru"]
})

df["City"] = df["City"].fillna(df["City"].mode()[0])

print(df)


        City
0  Bengaluru
1     Mysuru
2  Bengaluru
3  Bengaluru
4  Bengaluru


### Explanation

The code finds the most frequently occurring city and replaces the missing value with it.

Here, Bengaluru occurs most frequently, so the missing city is replaced with Bengaluru.

This is useful because we cannot calculate a mean or median for categories such as city, gender, product type, or department. Mode provides a simple way to handle missing categorical data before model training.


# 4. Weighted Mean

### Definition

Weighted mean is an average where different values are given different levels of importance called weights

### Real-life Example

Suppose a student's final result depends on:

- Assignment = 20%
- Test = 30%
- Final Exam = 50%

The final exam has more influence because it has a higher weight.

### Formula

$$
\text{Weighted Mean} =
\frac{\sum x_i w_i}{\sum w_i}
$$

Where:
- $x_i$ = value
- $w_i$ = weight

### Numerical Example

Scores = 80, 90, 70

Weights = 0.2, 0.3, 0.5

$$
\text{Weighted Mean}
=
\frac{(80 \times 0.2) + (90 \times 0.3) + (70 \times 0.5)}
{0.2 + 0.3 + 0.5}
$$

$$
\text{Weighted Mean}
=
\frac{16 + 27 + 35}{1}
= 78
$$

### Python Implementation


In [13]:
import numpy as np

scores = [80, 90, 70]
weights = [0.2, 0.3, 0.5]

weighted_mean = np.average(scores, weights=weights)

print("Weighted Mean:", weighted_mean)


Weighted Mean: 78.0


### AI/ML Example

Weighted mean can be used to combine predictions from multiple machine learning models.


In [15]:
model1_prediction = 80
model2_prediction = 90

final_prediction = (
    model1_prediction * 0.7 +
    model2_prediction * 0.3
)

print("Final Prediction:", final_prediction)


Final Prediction: 83.0


### Explanation

The first model receives a weight of 70%, while the second model receives 30%.

$$
\text{Final Prediction}
=
(80 \times 0.7) + (90 \times 0.3)
= 83
$$

This approach can be used in ensemble learning where predictions from several models are combined. A model considered more reliable can be assigned a larger weight, giving it more influence on the final prediction.


# 5. Geometric Mean

### Definition

Geometric mean is an average calculated by multiplying positive values together and taking the nth root of their product

It is useful for values that change multiplicatively, such as ratios and growth factors.

### Real-life Example

Suppose an investment grows at different rates over several years. Geometric mean can represent its average compounded growth

### Formula

$$
\text{Geometric Mean}
=
\sqrt[n]{x_1 \times x_2 \times \cdots \times x_n}
$$

Where $n$ is the number of observations.

### Numerical Example

Data: 2, 8

$$
\text{Geometric Mean}
=
\sqrt{2 \times 8}
$$

$$
\text{Geometric Mean}
=
\sqrt{16}
= 4
$$

### Python Implementation


In [17]:
from scipy.stats import gmean

data = [2, 8]

geometric_mean = gmean(data)

print("Geometric Mean:", geometric_mean)


Geometric Mean: 4.0


### AI/ML Example

Geometric mean can combine multiple positive performance scores when we want poor performance in one score to reduce the overall combined result.


In [19]:
from scipy.stats import gmean

scores = [0.8, 0.9, 0.7]

combined_score = gmean(scores)

print("Combined Score:", combined_score)


Combined Score: 0.7958114415792784


### Explanation

The code multiplies the three performance scores and takes their cube root.

$$
\text{Combined Score}
=
\sqrt[3]{0.8 \times 0.9 \times 0.7}
\approx 0.796
$$

The geometric mean is useful when combining multiplicative ratios or positive scores A low value in one measure pulls down the combined score, so it can help represent balanced performance across several measures.


# 6. Harmonic Mean

### Definition

Harmonic mean is an average calculated using the reciprocals of the values It is especially useful for rates and ratios

### Real-life Example

A car travels the same distance at 60 km/h and 40 km/h

The correct average speed for these equal-distance trips is found using the harmonic mean.

### Formula

$$
\text{Harmonic Mean}
=
\frac{n}
{\frac{1}{x_1}+\frac{1}{x_2}+\cdots+\frac{1}{x_n}}
$$

### Numerical Example

Speeds = 60 km/h and 40 km/h

$$
\text{Harmonic Mean}
=
\frac{2}
{\frac{1}{60}+\frac{1}{40}}
$$

$$
=
\frac{2}
{\frac{2}{120}+\frac{3}{120}}
$$

$$
=
\frac{2}{\frac{5}{120}}
= 48
$$

$$
\text{Harmonic Mean} = 48 \text{ km/h}
$$

### Python Implementation


In [21]:
from scipy.stats import hmean

speeds = [60, 40]

harmonic_mean = hmean(speeds)

print("Harmonic Mean:", harmonic_mean)


Harmonic Mean: 47.99999999999999


### AI/ML Example

A very important AI/ML example is the F1-score which uses the harmonic mean of Precision and Recall

$$
F1 =
2 \times
\frac{\text{Precision} \times \text{Recall}}
{\text{Precision} + \text{Recall}}
$$

Suppose:

$$
\text{Precision} = 0.80
$$

$$
\text{Recall} = 0.60
$$

Then:

$$
F1 =
2 \times
\frac{0.80 \times 0.60}
{0.80 + 0.60}
$$

$$
F1 =
\frac{0.96}{1.40}
\approx 0.686
$$


In [23]:
precision = 0.80
recall = 0.60

f1 = 2 * (precision * recall) / (precision + recall)

print("F1 Score:", f1)


F1 Score: 0.6857142857142857


### Explanation

The F1-score combines precision and recall into one value using the harmonic mean.

If precision is high but recall is low, or recall is high but precision is low, the F1-score becomes lower. This prevents one strong metric from hiding poor performance in the other.

It is especially useful in classification problems where both precision and recall are important, such as spam detection, fraud detection, and medical classification.
